event/src/unvalidated/builder.rs

```diff
@@ -0,0 +1,343 @@
+        DataBuilder {
+            id,
+            prev,
+            data,
+            sep: self.sep,
+            additional: self.additional,
+        }
+    }
+}
+
+/// Builder for unsigned unvalidated events
+pub struct InitBuilder {
+    sep: Option<String>,
+    controllers: Vec<String>,
+    additional: ValueMap,
+}
+
+impl Additional for InitBuilder {
+    fn with_additional(mut self, key: String, value: Value) -> Self {
+        self.additional.insert(key, value);
+        self
+    }
+}
+
+impl Controllers for InitBuilder {
+    fn with_controller(mut self, controller: String) -> Self {
+        self.controllers.push(controller);
+        self
+    }
+}
+
+impl Sep for InitBuilder {
+    fn with_sep(mut self, sep: String) -> Self {
+        self.sep = Some(sep);
+        self
+    }
+}
+
+impl InitBuilder {
+    /// Create unsigned init event without data
+    pub async fn build(self) -> anyhow::Result<init::Payload<()>> {
+        if self.controllers.is_empty() {
+            anyhow::bail!("controllers must not be empty");
+        }
+        let sep = self
+            .sep
+            .ok_or_else(|| anyhow::anyhow!("sep must be specified"))?;
+        let header = init::Header::new(self.controllers, sep, self.additional);
+        let payload = init::Payload::new(header, None);
+        Ok(payload)
+    }
+
+    /// Create builder for init event with data
+    pub fn with_data<T>(self, data: T) -> InitWithDataBuilder<T> {
+        InitWithDataBuilder {
+            data,
+            controllers: self.controllers,
+            sep: self.sep,
+            additional: self.additional,
+        }
+    }
+}
+
+/// Builder for init event with data
+pub struct InitWithDataBuilder<D> {
+    data: D,
+    controllers: Vec<String>,
+    sep: Option<String>,
+    additional: ValueMap,
+}
+
+impl<D> Additional for InitWithDataBuilder<D> {
+    fn with_additional(mut self, key: String, value: Value) -> Self {
+        self.additional.insert(key, value);
+        self
+    }
+}
+
+impl<D> Controllers for InitWithDataBuilder<D> {
+    fn with_controller(mut self, controller: String) -> Self {
+        self.controllers.push(controller);
+        self
+    }
+}
+
+impl<D> Sep for InitWithDataBuilder<D> {
+    fn with_sep(mut self, sep: String) -> Self {
+        self.sep = Some(sep);
+        self
+    }
+}
+
+impl<D> InitWithDataBuilder<D> {
+    /// Build unsigned init event with data
+    pub async fn build(self) -> anyhow::Result<init::Payload<D>> {
+        if self.controllers.is_empty() {
+            anyhow::bail!("controllers must not be empty");
+        }
+        let sep = self
+            .sep
+            .ok_or_else(|| anyhow::anyhow!("sep must be specified"))?;
+        let header = init::Header::new(self.controllers, sep, self.additional);
+        let payload = init::Payload::new(header, Some(self.data));
+        Ok(payload)
+    }
+}
+
+/// Builder for data event
+pub struct DataBuilder<T> {
+    id: Cid,
+    prev: Cid,
```

Comment:
its weird to me that this builder seems to intersperse fields that are part of the data and fields that are part of the header.  I feel like it would be better to keep those separate, and have a HeaderBuilder and DataBuilder that the EventBuilders use for each of the pieces so it remains explicit at the call site which information goes where instead of flattening it into one top-level structure.